### PR TITLE
feat(access): migrate ACCESS_PRIVATE to access collections

### LINKED ISSUE #4559: Throw exceptions on system failures rather than returning false (Trac #4559)
_Original ticket http://trac.elgg.org/ticket/4559 on 42423065-04-06 by trac user mrclay, assigned to unknown._

Elgg version: Github Master

Particularly low-level, non-API functions/methods (e.g. `row_to_elggmetadata`) which should not fail under normal conditions should throw instead of returning false/null. This would allow these functions to have fewer return types and remove the need for the caller to have code handling the error (which they usually don't anyway, most just pass the buck).

### LINKED ISSUE #7997: ElggEntity::create() should throw exceptions
`ElggEntity::create()` returns `false` instead of throwing an exception on container permissions failure. Calling methods such as `ElggGroup::create()` do not expect a falsey return and cause WSOD.

### LINKED ISSUE #8073: ACCESS_PRIVATE should be a real ACL
Much like ACCESS_FRIENDS and for the same reasons (namely inheritability): https://github.com/Elgg/Elgg/issues/5038

Each user would have an access collection containing only themselves.  This would be the access_id stored on anything private to them and anything that inherits that access.

### LINKED ISSUE #9033: FR: save:before event
Sometimes this would be preferable to the create/update events. Consider a plugin that wants to make sure a user's objects stay private. With update/create you're always reacting to the given data and have to re-save(), which calls the update event again.

save:before would basically let you alter attributes before the storage phase (and calling save() during the save:before event would register an error and be ignored).

Not sure it should let you cancel it though...

### LINKED ISSUE #11201: Failed update persists entity attributes
```php
$ia = elgg_set_ignore_access(false);
$user->language = 'de';
$user->save(); // false

echo $user->language; // de
echo get_entity($user->guid)->language; // en
```

The problem is that subsequent calls to other methods that do not require permissions checks could propagate these attributes to caches.

### PR DESCRIPTION
**Pull request by hypeJunction**
Migrates equivocal ACCESS_PRIVATE to an ACL-based system.
All users get an owned private access collection. Entities with magic
ACCESS_PRIVATE access id have been updated to owner's private ACL.
Adds helper functions for working with new ACL levels.
Makes transitioning from old access levels to new ones easier by adding
normalization during save operations.
Moves away from caching user's ACL membership and use IN operator in SQL
queries towards using subqueries providing for more scalability. Refs #3872
Reduces boilerplate around magic access level by introducing intermediary objects.
Adds API to filter collection icons.

Fixes #8073
Fixes #11201
Fixes #9033
Fixes #7997
Fixes #4559

### CODE REVIEW

File: engine/classes/Elgg/Upgrades/MigratePrivateACL.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function run(Result $result, $offset) {
+		
+		$db = elgg()->db;
+		
+		$users = elgg_get_entities([
```
Comment (hypeJunction) [inline engine/classes/Elgg/Upgrades/MigratePrivateACL.php]: Need to also migrate objects and groups that own private entities

**Comment by hypeJunction** [comment]
This would be good to have in 3.0, so we can do proper refactor of ACLs in 3.x series.

**Comment by jdalsem** [comment]
Unfortunately this can not go into 3.x as it is breaking

**Comment by hypeJunction** [comment]
I don't think current situation with access levels is acceptable. It's half baked.

**Comment by jdalsem** [comment]
that may be, but ACL's can be refactored during 3.x, but private ACL needs to stay because of BC... sorry

**Comment by jdalsem** [comment]
or think of a way to do this BC (setting ACCESS_PRIVATE and getting ACCESS_PRIVATE should work)

**Comment by jdalsem** [comment]
i can see you did some magic for the access levels, but this PR is really hard to examine as it is done on master and includes all kind of other commits

**Comment by jdalsem** [comment]
is it possible to split this into multiple PRs? As you fix 6 issues in one PR

**Comment by hypeJunction** [comment]
Everything is possible, but I am not going to spend more time on it.

**Comment by jdalsem** [comment]
> Everything is possible, but I am not going to spend more time on it.

well it is needed to get the PR in... and also to get rid of the RC issues.

**Comment by hypeJunction** [comment]
To get a PR in, all is needed is will and passing tests. Giving all the junk, it's a snowball effect. I try to fix one thing and there twenty other things in the way.
Whatever, I stop giving a flying fuck about Elgg core, it's exhausting.

**Pull request closed without merging**